source/common/network/udp_listener_impl.cc

```diff
@@ -112,56 +122,65 @@ void UdpListenerImpl::handleReadCallback() {
-                                 "{}, address length: {}",
-                                 addr.ss_family, local_address->asString(), recv_result.result_.rc_,
-                                 addr_len));
-      break;
+    try {
+      peer_address = Address::addressFromSockAddr(
+          addr, addr_len, local_address->ip()->version() == Address::IpVersion::v6);
+    } catch (const EnvoyException&) {
+      // Intentional no-op. The assert should fail below
     }
 
     RELEASE_ASSERT((peer_address != nullptr),
                    fmt::format("Unable to get remote address for fd: {}, local address: {} ",
                                socket_.ioHandle().fd(), local_address->asString()));
 
-    RELEASE_ASSERT((local_address != nullptr),
-                   fmt::format("Unable to get local address for fd: {}", socket_.ioHandle().fd()));
+    // Unix domain sockets are not supported
+    RELEASE_ASSERT(peer_address->type() == Address::Type::Ip,
+                   fmt::format("Unsupported peer address: {} local address: {}, receive size: "
+                               "{}, address length: {}",
+                               peer_address->asString(), local_address->asString(),
+                               recv_result.result_.rc_, addr_len));
 
-    cb_.onData(UdpData{local_address, peer_address, std::move(recv_result.buffer_)});
+    UdpRecvData recvData{local_address, peer_address, std::move(recv_result.buffer_)};
+    cb_.onData(recvData);
 
   } while (true);
 }
 
-void UdpListenerImpl::handleWriteCallback() { cb_.onWriteReady(socket_); }
+void UdpListenerImpl::handleWriteCallback() {
+  ENVOY_UDP_LOG(trace, "handleWriteCallback");
+  cb_.onWriteReady(socket_);
+}
+
+Event::Dispatcher& UdpListenerImpl::dispatcher() { return dispatcher_; }
+
+const Address::InstanceConstSharedPtr& UdpListenerImpl::localAddress() const {
+  return socket_.localAddress();
+}
+
+void UdpListenerImpl::send(const UdpSendData& send_data) {
+  ENVOY_UDP_LOG(trace, "send");
+  Buffer::Instance& buffer = send_data.buffer_;
+  uint64_t num_slices = buffer.getRawSlices(nullptr, 0);
+  STACK_ARRAY(slices, Buffer::RawSlice, num_slices);
+  buffer.getRawSlices(slices.begin(), num_slices);
+  Api::IoCallUint64Result send_result =
+      socket_.ioHandle().sendmsg(slices.begin(), num_slices, 0, *send_data.send_address_);
+
+  if (send_result.ok()) {
+    ENVOY_UDP_LOG(trace, "sendmsg sent:{} bytes", send_result.rc_);
+  } else {
+    if (send_result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
+      ENVOY_UDP_LOG(debug, "sendmsg dropped message");
+    } else {
+      ENVOY_UDP_LOG(debug, "sendmsg failed with error {}",
+                    static_cast<int>(send_result.err_->getErrorCode()));
+    }
+
+    // TODO(sumukhs) - Convert the OnError method to use IoErrors
+    cb_.onError(UdpListenerCallbacks::ErrorCode::SyscallError,
+                static_cast<int>(send_result.err_->getErrorCode()));
+  }
```

Comment:
Per my other comment I think you want to drain the buffer here?

Reply:
Is it possible that even in the success case, the number of bytes sent is lesser than requested size? in that case, do we need to drain as well?

Also - what is the purpose of draining the buffer by calling the "drain" API on the buffer interface? It frees up the memory which is expected to be done by the callee?

Reply:
I did some literature reading - looks like sendmsg can return with fewer bytes than requested only in the case of blocking call. Since the code always uses async  IO, this shouldn't happen.

I am still confused on who owns the responsibility of draining buffers under various conditions. For example, the BufferImpl::write() method takes in a network io handle and calls write and then drain only in the success case. In the failure case, the method returns the failure code.

Analogous to the above, I would have expected the udp listener to drain the buffer in the happy path (is the general rule of thumb for the component that invokes an API on the Network::io_handle to take charge of draining the buffer if needed?). However, since the udp listener send() does not return an error code and is best effort, it must drain the buffers in the error paths too? 

Is this the right way to think about this?

Reply:
In terms of who is responsible for draining buffers, it's really just an API contract which I admit we are not great about documenting. The TCP connection impl will actually move the writter buffer data into a different TCP write buffer. IMO in order to be consistent, we should drain the UDP buffer after sending it to the OS, at least in the case where we send partial data. If no data is sent (for example because the size is larger than discovered path MTU) maybe we leave it undrained in case the caller wants to fragment it? WDYT? Either way I think we need to document and think about the behavior here.

Reply:
Thanks for clarifying. I have currently pushed an update where I am draining the buffer in both the cases. We could add the logic to check the error code and keep it undrained when we expose the MTU information so that the caller can break it up?

Reply:
I guess my feeling is that we should drain any data that is sent, but leave any data that is not sent? Then the caller can decide what to do? Also, I think we need to return error information from this function so it should't be void? That would allow the caller to understand what is going on and adapt?

Reply:
Yes, thats a reasonable approach. 

In the design docs, I had thought of treating the udp send as "best effort" - as in packets could be lost. So the first iteration does not handle "writeready" socket events and along the same thinking, I had kept the return type void  for "send" as the sender cannot assume the data was successfully sent even if the packets leave the machine.

I will add the return code in the next iteration.

Reply:
@mattklein123 - pushed an update last night. Let me know if any other changes are required.

Reply:
@danzh2010  - This is the previous conversation on the drain topic.
@mattklein123  suggested that we do not drain any buffers that is not sent. Can we reach consensus on what is the recommended approach here?

Reply:
UDP sendmsg() doesn't support partial send. It sends all or nothing. If we want to drain buffer here, for sure we can drain the whole buffer when it returns OK. If it returns any error, including Again, nothing actually has been sent. In this case, I think probably we want to let the caller decide.

Reply:
I see... thanks for the information @danzh2010. Does the current iteration of the code look fine then? In the error case, we will do a drain(0) which should keep the buffer as it is right?

Reply:
Yes. Looks good to me.